astropy/table/column.py

```diff
@@ -1191,12 +1191,21 @@ def __new__(cls, data=None, name=None, mask=None, fill_value=None,
                 copy=False, copy_indices=True):
 
         if mask is None:
-            # Issue #7399 with fix #7422.  Passing mask=None to ma.MaskedArray
-            # is extremely slow (~3 seconds for 1e7 elements), while mask=False
-            # gets quickly broadcast to the expected bool array of False.
-            mask = getattr(data, 'mask', False)
-            if mask is not False:
-                mask = np.array(mask, copy=copy)
+            # If mask is None then we need to determine the mask (if any) from the data.
+            # The naive method is looking for a mask attribute on data, but this can fail,
+            # see #8816.  Instead use ``MaskedArray`` to do the work.
+            data_ma = ma.MaskedArray(data)
+            mask = data_ma.mask.copy() if copy else data_ma.mask
```

Comment:
Maybe deal with the `np.bool_` case, which is just `nomask` here as well (was thinking of falling through but we don't want the `deepcopy`):
```
mask = data_ma.mask
if mask is np.ma.nomask:
    mask = False
elif copy:
    mask = mask.copy()
```

Reply:
Actually, could make first line:
```
mask = ma.MaskedArray(data).mask
```

Reply:
Awesome idea, done.  (And I came up with the `mask = ma.MaskedArray(data).mask` independently once I saw the code!)